crates/uplc/src/machine.rs

```diff
@@ -0,0 +1,407 @@
+            .get::<usize>(usize::from(name.index) - 1)
+            .cloned()
+            .ok_or(Error::OpenTermEvaluated(Term::Var(name)))
+    }
+
+    fn step_and_maybe_spend(&mut self, step: StepKind) -> Result<(), Error> {
+        let index = step as u8;
+        self.unbudgeted_steps[index as usize] += 1;
+        self.unbudgeted_steps[7] += 1;
+
+        if self.unbudgeted_steps[7] >= self.slippage {
+            self.spend_unbudgeted_steps()?;
+        }
+
+        Ok(())
+    }
+
+    fn spend_unbudgeted_steps(&mut self) -> Result<(), Error> {
+        for i in 0..self.unbudgeted_steps.len() - 1 {
+            let mut unspent_step_budget =
+                self.costs.machine_costs.get(StepKind::try_from(i as u8)?);
+
+            unspent_step_budget.occurences(self.unbudgeted_steps[i] as i64);
+
+            self.spend_budget(unspent_step_budget)?;
+
+            self.unbudgeted_steps[i] = 0;
+        }
+
+        self.unbudgeted_steps[7] = 0;
+
+        Ok(())
+    }
+
+    fn spend_budget(&mut self, spend_budget: ExBudget) -> Result<(), Error> {
+        self.ex_budget.mem -= spend_budget.mem;
+        self.ex_budget.cpu -= spend_budget.cpu;
+
+        if self.ex_budget.mem < 0 || self.ex_budget.cpu < 0 {
+            Err(Error::OutOfExError(self.ex_budget))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+#[derive(Clone)]
+enum Context {
+    FrameApplyFun(Value),
+    FrameApplyArg(Term<NamedDeBruijn>),
+    FrameForce,
+    NoFrame,
+}
+
+#[derive(Clone, Debug)]
+pub enum Value {
+    Con(Constant),
+    Delay(Term<NamedDeBruijn>),
+    Lambda {
+        parameter_name: NamedDeBruijn,
+        body: Term<NamedDeBruijn>,
+    },
+    Builtin {
+        fun: DefaultFunction,
+        term: Term<NamedDeBruijn>,
+        runtime: BuiltinRuntime,
+    },
+}
+
+impl Value {
+    pub fn is_integer(&self) -> bool {
+        matches!(self, Value::Con(Constant::Integer(_)))
+    }
+
+    pub fn is_bool(&self) -> bool {
+        matches!(self, Value::Con(Constant::Bool(_)))
+    }
+
+    pub fn to_ex_mem(&self) -> i64 {
+        match self {
+            Value::Con(c) => match c {
+                Constant::Integer(i) => {
+                    if *i == 0 {
+                        1
+                    } else {
+                        ((i.abs() as f64).log2().floor() as i64 / 64) + 1
+                    }
+                }
+                Constant::ByteString(b) => (((b.len() - 1) / 8) + 1) as i64,
+                Constant::String(s) => s.chars().count() as i64,
+                Constant::Unit => 1,
+                Constant::Bool(_) => 1,
+            },
+            Value::Delay(_) => 1,
+            Value::Lambda { .. } => 1,
+            Value::Builtin { .. } => 1,
+        }
+    }
+
+    pub fn expect_type(&self, r#type: Type) -> Result<(), Error> {
+        match self {
+            Value::Con(constant) => {
+                let constant_type = Type::from(constant);
+
+                if constant_type == r#type {
```

Comment:
I've never seen this before. I assume it's essentially the same as `matches!()`?

Reply:
the r# thing?

Reply:
Yeah. You're just checking the variants match?